c10/cuda/CUDAStream.cpp

```diff
@@ -321,6 +363,24 @@ cudaStream_t CUDAStream::stream() const {
   return ptr->stream;
 }
 
+CUDAAssert* CUDAStream::assert_state() const {
+  auto ptr = CUDAStream_internals(*this);
+  AT_ASSERT(ptr);
+
+  // lazy create assert state (TODO: support concurrent calls)
+  if (ptr->assert_state == nullptr) {
+    // switch to device to associate host memory with it
+    CUDAGuard device_guard(device_index());
+
+    // allocate assert memory
+    C10_CUDA_CHECK(cudaHostAlloc(
+      (void **)&ptr->assert_state, sizeof(ptr->assert_state),
+      cudaHostAllocMapped/* | cudaHostAllocPortable*/));
+  }
```

Comment:
You could do a check for nullptr and opportunistically allocate followed by a CAS and potential free.